### ISSUE TITLE
msgfmt is not overridable in cross [binaries]

### ISSUE THREAD

**Issue by hughsie** [mesonbuild/meson#10376]
**Describe the bug**

Setting `msgfmt = '/usr/x86_64-w64-mingw32/sys-root/mingw/bin/msgfmt.exe'` in the `[binaries]` section seems to have no effect. Overriding other binaries like `glib-mkenums` seems to work fine.

**To Reproduce**

    i18n.gettext(meson.project_name(),
      preset : 'glib',
      args: [
      '--default-domain=' + meson.project_name(),
      ]
    )

**Expected behavior**

    wine /usr/x86_64-w64-mingw32/sys-root/mingw/bin/msgfmt.exe ../po/pl.po -o po/pl/LC_MESSAGES/fwupd.mo

**Actual behavior**

    msgfmt ../po/pl.po -o po/pl/LC_MESSAGES/fwupd.mo

**system parameters**
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.)

Fedora 36.

* what Python version are you using e.g. 3.8.0

3.10.4

* what `meson --version`

0.60.3

* what `ninja --version` if it's a Ninja build

1.10.2

**Comment by hughsie** [comment]
A potentially similar issue is that meson tries to run `mono` for this override:

	[binaries]
	glib-compile-resources = '/usr/x86_64-w64-mingw32/sys-root/mingw/bin/glib-compile-resources.exe'
	exe_wrapper = '/usr/bin/wine'

	[properties]
	root = '/usr/x86_64-w64-mingw32/sys-root/mingw'
	needs_exe_wrapper = true

	[host_machine]
	system = 'windows'
	cpu_family = 'x86_64'
	cpu = 'x86_64'
	endian = 'little'

e.g.

    mono /usr/x86_64-w64-mingw32/sys-root/mingw/bin/glib-compile-resources.exe ../src/fwupd.gresource.xml --sourcedir ../src/. --sourcedir ../src --c-name fu --internal --generate --target src/fwupd-resources.h

it seems "internal" things don't seem to be wired up in the same way as if you used `find_program` in a `meson.build`...

**Comment by hughsie** [comment]
With master I still get:

    ../data/remotes.d/meson.build:20:7: ERROR: Program 'msgfmt' not found or not executable

where that's just a call to `i18n.merge_file` -- and I've got with `msgfmt = '/usr/x86_64-w64-mingw32/sys-root/mingw/bin/msgfmt.exe'` set in the cross file -- which certainly exists and is executable.

**Comment by anarazel** [comment]
The reason this doesn't work is presumably that msgfmt is looked up as build machine tool:
https://github.com/mesonbuild/meson/blob/b1abda2b4b9ea85551abe073b55092848285ccca/mesonbuild/modules/i18n.py#L228
so overriding this in the cross file won't work?

**Comment by eli-schwartz** [comment]
That's correct, looking it up as a cross-compiled tool would mean that people who have a native version installed would still have to build and run a wine version which seems suboptimal since this produces machine-independent compiled output catalogs.

**Comment by hughsie** [comment]
> people who have a native version installed would still have to build and run a wine version

In my instance, I have the mingw tools installed, but don't have the native version installed, e.g. in this case we need to have *both* installed for the build to succeed: https://github.com/fwupd/fwupd/blob/main/contrib/ci/dependencies.xml#L494

**Comment by xclaesse** [comment]
> Setting `msgfmt = '/usr/x86_64-w64-mingw32/sys-root/mingw/bin/msgfmt.exe'` in the `[binaries]` section seems to have no effect.

Maybe try `msgfmt = ['wine', '/usr/x86_64-w64-mingw32/sys-root/mingw/bin/msgfmt.exe']`, does that work?

**Comment by hughsie** [comment]
> Maybe try msgfmt = ['wine',

Nope, no effect whatsoever.

**Comment by xclaesse** [comment]
Please double check you're using meson master, and check  your logs to see what command it runs for msgfmt, because it seems to work for me. Note sure about your exact setup, but defining a binary as `msgfmt = ['wine', '/usr/x86_64-w64-mingw32/sys-root/mingw/bin/msgfmt.exe']` and then using find_program() for it, it will run it with wine.

**Comment by hughsie** [comment]
I agree, this works really well:

    msgfmt = find_program('msgfmt')
    error(msgfmt.full_path())

which gives the expected:

    Program msgfmt found: YES
    ../meson.build:199:0: ERROR: Problem encountered: /usr/x86_64-w64-mingw32/sys-root/mingw/bin/msgfmt.exe

... but this isn't what `i18n.merge_file` seems to be using. i.e. `i18n.merge_file` seems to want '/usr/bin/msgfmt' but `find_program` does the right thing.

**Comment by anarazel** [comment]
@hughsie Does it still work if you do `find_program('msgfmt', native: true)`? Are you specifying the msgfmt path in the native rather than the cross file?

**Comment by hughsie** [comment]
> Does it still work if you do find_program('msgfmt', native: true)? 

Ohh, interesting. Using:

    warning('native:true: @0@'.format(find_program('msgfmt', native: true).full_path()))
    warning('native:false: @0@'.format(find_program('msgfmt', native: false).full_path()))

I see:

../meson.build:198: WARNING: native:true: /usr/bin/msgfmt
../meson.build:199: WARNING: native:false: /usr/x86_64-w64-mingw32/sys-root/mingw/bin/msgfmt.exe

> Are you specifying the msgfmt path in the native rather than the cross file?

I'm using a cross file that has:

    [binaries]
    exe_wrapper = '/usr/bin/wine'
    windmc = '/usr/bin/x86_64-w64-mingw32-windmc'
    msgfmt = ['wine', '/usr/x86_64-w64-mingw32/sys-root/mingw/bin/msgfmt.exe']

To my naive eyes it seems that `i18n.merge_file` is using the native:true value, not the native:false value.

> Would help if you can share the docker image you're using with more detailed steps to reproduce

This is what I've done; I can do a container tomorrow if that would be easier. To test this I've been doing:

    git clone https://github.com/fwupd/fwupd.git
    ./contrib/setup # and install the native deps when prompted
    sudo ./contrib/ci/fwupd_setup_helpers.py -v mingw64 install-dependencies # only works on Fedora, but you can get the list of deps with `./contrib/ci/fwupd_setup_helpers.py -v mingw64 -o fedora get-dependencies` for other OSs.
    # manually remove the glib2-dev package for your distro, or do something like `sudo /usr/bin/msgfmt /usr/bin/msgfmt.bak`
    add `msgfmt = ['wine', '/usr/x86_64-w64-mingw32/sys-root/mingw/bin/msgfmt.exe']` to the `[binaries]` section of `contrib/mingw64.cross`
    # install meson master systemwide, or patch ./contrib/build-windows.sh to call ./path-to/meson.py
    ./contrib/build-windows.sh

> Seems you've got many issues related to wine and cross compilation, you're doing that for fwup ci, right?

Correct.

**Comment by eli-schwartz** [comment]
> To my naive eyes it seems that `i18n.merge_file` is using the native:true value, not the native:false value.

This was mentioned above:

> The reason this doesn't work is presumably that msgfmt is looked up as build machine tool:
https://github.com/mesonbuild/meson/blob/b1abda2b4b9ea85551abe073b55092848285ccca/mesonbuild/modules/i18n.py#L228
so overriding this in the cross file won't work?

To which I responded:

>That's correct, looking it up as a cross-compiled tool would mean that people who have a native version installed would still have to build and run a wine version which seems suboptimal since this produces machine-independent compiled output catalogs.

**Comment by hughsie** [comment]
Right, but might it make sense to look for it as a native tool, and if not found, look it up as a cross-built tool? If it's truly machine-independent then that means it's completely safe to use from either. At the moment I'm installing ~5MB of native packages to run a tool that already exists as a cross-built binary.

**Comment by eli-schwartz** [comment]
Perhaps we could add a `native: 'either'`, then internally use that for programs where either one is fine.

**Comment by anarazel** [comment]
@hughsie 
> Right, but might it make sense to look for it as a native tool, and if not found, look it up as a cross-built tool? If it's truly machine-independent then that means it's completely safe to use from either. At the moment I'm installing ~5MB of native packages to run a tool that already exists as a cross-built binary.

Yes, I was thinking that might be a good idea. An 'either' value for platform-indep tools makes a lot of sense imo. Often a native tool will be faster, but sometimes it'll be too hard to install. Neither meson.build nor modules should need explicit code for something like that imo.

But still would be good to confirm that additionally specifying a native file with the wine /path/to/msgfmt works?

**Comment by anarazel** [comment]
@hughsie just like you can specify --cross-file you can also specify a "native" file with --native-file. If you set msgfmt to `wine /path/to/mingw/msgfmt` in that, it should redirect it, from what I can see.

**Comment by hughsie** [comment]
TIL!

Using this:

    --cross-file=../contrib/mingw64.cross \
    --native-file=../contrib/mingw64.cross \

 I get this output...

    ../meson.build:198: WARNING: native:true: /usr/x86_64-w64-mingw32/sys-root/mingw/bin/msgfmt.exe
    ../meson.build:199: WARNING: native:false: /usr/x86_64-w64-mingw32/sys-root/mingw/bin/msgfmt.exe

i.e. --native-file overrides the native:true case.

So, is this bug wasting everybodies time not knowing about `--native-file` or should `native: 'either'` be used in `i18n.merge_file` instead?

**Comment by eli-schwartz** [comment]
It seems like there's a genuine use case for `native: 'either'`, so it was a useful discussion...

Manually hardcoding the exe_wrapper as the first component of a cross binary in the native file is genuinely awkward.

**Comment by anarazel** [comment]
Agreed. Separately: I wonder if it's worth adding a warning message, at least to the log, if a native tool isn't found but there's a cross binary or vice versa.

**Comment by hughsie** [comment]
I think if it's really immaterial which one is used (e.g. msgfmt) a warning seems overkill, but it's certainly something I can ignore if needed :)
